### ISSUE TITLE
Expanding Low-Density Latent Regions for Open-Set Object Detection

### ISSUE THREAD

**Issue by zhuoxiao-chen** [zhuoxiao-chen/please_read_paper_every_day#32]
### Paper Link
https://ieeexplore.ieee.org/document/9880281

### Code Link
https://github.com/csuhan/opendet2

## Motivation
<img width="600" alt="image" src="https://user-images.githubusercontent.com/50613939/232643847-f29e016a-dbd9-4ead-b5bf-47ac09b23079.png">

- Traditional threshold-based methods only maintain limited low-density regions, which cannot cover all unknown objects. 
- Goal is to encourage compact features of known classes, leaving more low-density regions for unknown classes. 

## Methodology
<img width="1200" alt="image" src="https://user-images.githubusercontent.com/50613939/232670625-2af70cb0-a381-4a39-980c-f9622fe18a58.png">

#### Class-balanced memory bank
- For each class $c \in C_{\mathcal{K}}$, initialize a memory bank $M(c)$
- Sample proposals with IoU $> T_m$ 
- For each mini-batch, sample $q$ proposals that are least similar (_i.e._, minimum cosine similarity) with existing exemplars in  $M(c)$

#### Instance-Level Contrastive Learning
An Instance Contrastive (IC) Loss to learn more compact features of object proposals.

$$
\mathcal{L}_{I C}\left(\mathbf{z}_i\right)=\frac{1}{\left|M\left(\mathbf{c}_i\right)\right|} \sum_{\mathbf{z}_j \in M\left(\mathbf{c}_i\right)} \log \frac{\exp \left(\mathbf{z}_i \cdot \mathbf{z}_j / \tau\right)}{\sum_{\mathbf{z}_k \in A\left(c_i\right)} \exp \left(\mathbf{z}_i \cdot \mathbf{z}_k / \tau\right)}
$$

$$
A\left(c_i\right)=M \backslash M\left(c_i\right)
$$

### Unknown Probability Learner (UPL)

- Augment the K-way classifier with the K+1-way classifier, where K+1 denotes the unknown class
- Cross-Entropy (CE) Loss, if $c=c^*$ then $y_c=1$, else, $y_c=0$ 

$$
p_c=\operatorname{softmax}\left(s_c\right)=\frac{\exp \left(s_c\right)}{\sum_{j \in C} \exp \left(s_j\right)}
$$

$$
\mathcal{L}_{C E}=-\sum_{c \in C} y_c \log \left(p_c\right)
$$

- Unknown Probability (UP) Loss: jointly optimizing the $L_{UP}$ and $L_{CE}$ divides more low density latent regions (in gray color) from known classes

$$
p_u^{\prime}=\frac{\exp \left(s_u\right)}{\sum_{j \in C, j \neq c^*} \exp \left(s_j\right)}
$$

$$
\mathcal{L}_{U P}=-\log \left(p_u^{\prime}\right)
$$

> Optimizing $L_{CE}$ enables the probability of GT to be higher, all other classes to be lower. Optimizing $L_{UP}$ enables the probability of UNK to be higher, all other classes except GT to be lower. Which means $L_{UP}$ increase the probability of UNK, while $L_{CE}$ increase the probability of GT and decrease the probability of UNK.  More certain, higher probability of GT and lower probability of UNK. Higher probability of UNK means lower certainty. 

- Uncertainty guided hard example mining to optimize $L_{UP}$: top-k entropy or top-k min max-probability.

### Evaluation Metrics
**Wilderness Impact (WI)**: $W I=\left(\frac{P_{\mathcal{K}}}{P_{\mathcal{K} \cup \mathcal{U}}}-1\right) \times 100$, where $P$ denote the precision under a recall level of $0.8$.

**Absolute Open-Set Error (AOSE)**  counts the number of misclassified unknown objects.

**mean Average Precision (mAP) of known classes**: ($\text{mAP}_\mathcal{K}$)

**Average Precision (mAP) of UNK classes**: ($\text{mAP}_\mathcal{U}$)

### Baselines 
- **Faster R-CNN (FR-CNN)**
- **Dropout Sampling (DS)**: Dropout sampling for robust object detection in open-set conditions
- **ORE**: Towards open world object detection.
- **PROSER**: Learning placeholders for open-set recognition.

## Experiments 
<img width="1200" alt="image" src="https://user-images.githubusercontent.com/50613939/232716020-535151b2-8279-427b-b11c-3f073023cd7d.png">

<img width="400" alt="image" src="https://user-images.githubusercontent.com/50613939/232718807-9945a573-9ad4-4333-835e-6317a713c37c.png">
